### PR TITLE
Add reflection invoke benchmarks

### PR DESCRIPTION
**Pull request by steveharter**
Adds the first invoke-based reflection benchmarks. In 7.0 there was various refactoring and support for generated IL for the invoke, which significantly improved performance of most scenarios.

Here's the 6.0 vs 7.0 results:
```
|                                               Method |        Job |                                                                                                   Toolchain |       Mean |     Error |     StdDev |     Median |        Min |        Max | Ratio | RatioSD |  Gen 0 | Allocated | Alloc Ratio |
|----------------------------------------------------- |----------- |------------------------------------------------------------------------------------------------------------ |-----------:|----------:|-----------:|-----------:|-----------:|-----------:|------:|--------:|-------:|----------:|------------:|
|                                      Method0_NoParms | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe |  35.989 ns | 0.7340 ns |  0.7209 ns |  35.797 ns |  35.229 ns |  37.316 ns |  4.57 |    0.09 |      - |         - |          NA |
|                                      Method0_NoParms | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe |   7.926 ns | 0.0546 ns |  0.0427 ns |   7.926 ns |   7.868 ns |   7.991 ns |  1.00 |    0.00 |      - |         - |          NA |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
| StaticMethod4_arrayNotCached_int_string_struct_class | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe | 150.434 ns | 1.1955 ns |  1.1183 ns | 150.566 ns | 148.512 ns | 151.906 ns |  2.43 |    0.05 | 0.0096 |     104 B |        1.00 |
| StaticMethod4_arrayNotCached_int_string_struct_class | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe |  62.006 ns | 1.0492 ns |  0.9814 ns |  62.344 ns |  60.016 ns |  63.263 ns |  1.00 |    0.00 | 0.0097 |     104 B |        1.00 |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
|                StaticMethod4_int_string_struct_class | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe | 128.163 ns | 1.3783 ns |  1.2893 ns | 127.733 ns | 126.911 ns | 130.640 ns |  3.11 |    0.05 |      - |         - |          NA |
|                StaticMethod4_int_string_struct_class | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe |  41.238 ns | 0.6350 ns |  0.5940 ns |  41.190 ns |  40.473 ns |  42.205 ns |  1.00 |    0.00 |      - |         - |          NA |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
|    StaticMethod4_ByRefParams_int_string_struct_class | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe | 283.771 ns | 1.2647 ns |  1.1830 ns | 283.549 ns | 282.303 ns | 285.694 ns |  1.71 |    0.02 | 0.0045 |      48 B |        1.00 |
|    StaticMethod4_ByRefParams_int_string_struct_class | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe | 166.303 ns | 2.4653 ns |  2.3060 ns | 166.612 ns | 163.781 ns | 171.514 ns |  1.00 |    0.00 | 0.0040 |      48 B |        1.00 |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
|                                       Ctor0_NoParams | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe |  50.415 ns | 0.2689 ns |  0.2384 ns |  50.431 ns |  49.987 ns |  50.924 ns |  4.93 |    0.04 | 0.0052 |      56 B |        1.00 |
|                                       Ctor0_NoParams | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe |  10.234 ns | 0.0610 ns |  0.0541 ns |  10.243 ns |  10.115 ns |  10.322 ns |  1.00 |    0.00 | 0.0053 |      56 B |        1.00 |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
|               Ctor0_ActivatorCreateInstance_NoParams | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe |   7.929 ns | 0.1021 ns |  0.0905 ns |   7.905 ns |   7.847 ns |   8.139 ns |  1.04 |    0.02 | 0.0053 |      56 B |        1.00 |
|               Ctor0_ActivatorCreateInstance_NoParams | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe |   7.600 ns | 0.1369 ns |  0.1280 ns |   7.585 ns |   7.428 ns |   7.816 ns |  1.00 |    0.00 | 0.0054 |      56 B |        1.00 |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
|                        Ctor4_int_string_struct_class | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe | 148.155 ns | 1.5491 ns |  1.3732 ns | 147.567 ns | 146.829 ns | 150.935 ns |  3.32 |    0.05 | 0.0048 |      56 B |        1.00 |
|                        Ctor4_int_string_struct_class | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe |  44.622 ns | 0.4233 ns |  0.3960 ns |  44.672 ns |  43.530 ns |  45.130 ns |  1.00 |    0.00 | 0.0053 |      56 B |        1.00 |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
|                        Ctor4_ActivatorCreateInstance | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe | 409.933 ns | 9.0823 ns | 10.4592 ns | 414.259 ns | 386.667 ns | 420.944 ns |  1.54 |    0.04 | 0.0418 |     448 B |        1.08 |
|                        Ctor4_ActivatorCreateInstance | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe | 264.523 ns | 2.1209 ns |  1.9839 ns | 264.031 ns | 262.430 ns | 268.461 ns |  1.00 |    0.00 | 0.0389 |     416 B |        1.00 |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
|                                     Property_Get_int | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe |  54.411 ns | 0.2750 ns |  0.2438 ns |  54.313 ns |  54.140 ns |  54.868 ns |  4.91 |    0.05 | 0.0022 |      24 B |        1.00 |
|                                     Property_Get_int | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe |  11.078 ns | 0.1349 ns |  0.1196 ns |  11.041 ns |  10.919 ns |  11.338 ns |  1.00 |    0.00 | 0.0023 |      24 B |        1.00 |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
|                                   Property_Get_class | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe |  37.038 ns | 0.1706 ns |  0.1425 ns |  36.992 ns |  36.859 ns |  37.319 ns |  4.52 |    0.04 |      - |         - |          NA |
|                                   Property_Get_class | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe |   8.186 ns | 0.0767 ns |  0.0680 ns |   8.160 ns |   8.114 ns |   8.355 ns |  1.00 |    0.00 |      - |         - |          NA |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
|                                     Property_Set_int | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe |  68.409 ns | 1.3855 ns |  1.2960 ns |  68.015 ns |  67.161 ns |  71.427 ns |  2.41 |    0.07 | 0.0022 |      24 B |        1.00 |
|                                     Property_Set_int | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe |  28.171 ns | 0.5965 ns |  0.6630 ns |  28.334 ns |  27.322 ns |  29.424 ns |  1.00 |    0.00 | 0.0022 |      24 B |        1.00 |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
|                                   Property_Set_class | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe |  61.589 ns | 0.2841 ns |  0.2519 ns |  61.550 ns |  61.272 ns |  61.920 ns |  2.82 |    0.02 |      - |         - |          NA |
|                                   Property_Set_class | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe |  21.806 ns | 0.1523 ns |  0.1425 ns |  21.765 ns |  21.660 ns |  22.062 ns |  1.00 |    0.00 |      - |         - |          NA |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
|                                        Field_Get_int | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe |  30.993 ns | 0.1622 ns |  0.1438 ns |  30.972 ns |  30.768 ns |  31.259 ns |  0.87 |    0.01 | 0.0022 |      24 B |        1.00 |
|                                        Field_Get_int | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe |  35.749 ns | 0.2806 ns |  0.2487 ns |  35.752 ns |  35.461 ns |  36.344 ns |  1.00 |    0.00 | 0.0021 |      24 B |        1.00 |
|                                                      |            |                                                                                                             |            |           |            |            |            |            |       |         |        |           |             |
|                                        Field_Set_int | Job-BQDNCA | \runtime60\artifacts\bin\testhost\net6.0-windows-Release-x64\shared\Microsoft.NETCore.App\6.0.9\corerun.exe |  36.037 ns | 0.2146 ns |  0.1903 ns |  36.018 ns |  35.744 ns |  36.418 ns |  0.98 |    0.01 | 0.0023 |      24 B |        1.00 |
|                                        Field_Set_int | Job-RQCXYM |   \runtime\artifacts\bin\testhost\net7.0-windows-Release-x64\shared\Microsoft.NETCore.App\7.0.0\corerun.exe |  36.766 ns | 0.2736 ns |  0.2425 ns |  36.666 ns |  36.378 ns |  37.268 ns |  1.00 |    0.00 | 0.0023 |      24 B |        1.00 |
```

### CODE REVIEW

File: src/benchmarks/micro/runtime/System.Reflection/Invoke.cs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        [Benchmark]
+        public void Ctor0_ActivatorCreateInstance_NoParams()
+        {
+            Activator.CreateInstance(typeof(MyClass));
+        }
+
+        [Benchmark]
+        public void Ctor4_int_string_struct_class()
+        {
+            object[] args = new object[] { 42, "Hello", default(MyBlittableStruct), s_MyClass };
```
Comment (danmoseley) [inline src/benchmarks/micro/runtime/System.Reflection/Invoke.cs:130]: Would it be better to pull creating these arrays out of the benchmarks? They could presumably simply be reused.

Doesn't matter much as presumably it's very cheap but it'd be consistent.

Comment (steveharter) [reply src/benchmarks/micro/runtime/System.Reflection/Invoke.cs:130]: The benchmarks with "PreInstantiatedObjectArray" use a cached array. The rest (majority) allocate the array.

 I suppose I could flip it, and by default cache most.

**Review by buyaa-n — APPROVED** [review]

File: src/benchmarks/micro/runtime/System.Reflection/Invoke.cs (hunk 1 above)
Comment (danmoseley) [reply src/benchmarks/micro/runtime/System.Reflection/Invoke.cs:130]: Why allocate each iteration for any of them though?

Comment (steveharter) [reply src/benchmarks/micro/runtime/System.Reflection/Invoke.cs:130]: It is measuring the actual real-world scenario where typically the arrays are not pre-allocated or pre-filled. Thus, these benchmarks are a fair, realistic benchmark especially when claiming that reflection is "X" times faster. This is also useful for when comparing against the proposed new reflection APIs being introduced in 8.0 which avoid the array allocation.

Given that, I do think I should flip the benchmarks so the default is pre-allocated (but not pre-filled) and then label the ones that allocate as "scenario"-based.

**Pull request merged**